### ISSUE TITLE
Log all hits

### ISSUE THREAD

**Issue by JDGrimes** [WordPoints/hooks-api#55]
## Background

After some work on #53 and #10, I'm wondering whether we might be able to solve a range of problems by logging every fire, or every hit, to the database.
### Points Reactor

This would help the points reactor by providing a way for us to determine whether to actually award points for a hit, by calculating the number of hits for that entity and the number of reverses. See #53.
### Periods Extension

This would help the periods extension in a variety of ways, for both absolute and relative periods. For absolute periods, knowing the last time that an event fired (hit actually) is important for properly adjusting things when the time length for a period is increased or decreased. It is also important to know the time of the last hit for relative periods, since period length is calculated from the last hit.

It is also important that we be able to identify whether the original even occurred during the current period, when reversals occur.
## Questions

This seems to me to be a good idea, because it solves a variety of problems in one fell swoop. However, there are definitely some open questions that we'll need to explore.
### Retroactive API

How would this interact with the retroactive API?

In some ways, this API would be doing the same thing as the retroactive API; that is, it would be solving some issues that I was planning for the retroactive API to one day solve. However, I'm not really sure that the retroactive API was really the best solution for those things anyway, so I think it is a good thing for us to entertain this alternative possibility.

One example of where we would maybe use the retroactive API is in helping to handle edge-cases when the length of a period changes. The problem with that idea is that while the retroactive API can tell us whether an event has taken place, it can't tell us _when_ that event has taken place. So the only way that we would find it useful is if we performed the query each time a period started. That would be expensive, especially since these queries are uncached and can become rather complex. So this solution is much better.

The query API would not be helpful for the points reactor either. What we need to know there is not just whether an event has happened, but whether a hit was made, and the retroactive/query API cannot help us there. Also, the status of a post, for example, will have already been updated in the database before the hit occurs, so running a query at that point is useless.

I question whether even the hypothetical badges reactor would benefit from the query API. I suppose it would depend upon whether we make badges stateful or cumulative. If we make them stateful then the query API would be invaluable. But we'd need something like this proposal if we wanted to make them cumulative (or stateful with a one-way valve, I think).
### Fires, Hits, or Misses

Another important question is whether we would want to log all fires including misses, or only hits.

Logging misses, you understand, will cause the logs to accumulate at a rapid pace. So it is something that we'd like to avoid, if possible. However, I'm wondering if there may be some times where we need to log misses. So let's consider our use cases.

For the points reactor there is no need to log misses, only hits.

For the periods extension it is the same story. One thing that we need to know is whether past hits happened during the current period when a period's length is increased. When reversals occur, we need to know whether the original event that is being reversed actually caused a hit. To know that we'd just need to be able to check for a matching hit for that event in the database; if we didn't find one we could safely assume it was a miss then. So there would be no need to have the misses themselves logged for that purpose. However, I'm wondering whether it really is sufficient when, for example, a post's status is toggled multiple times. With the points reactor it doesn't matter, because it will not be reversing that event (it only listens for spam and posts don't have a spam status by default), but what about other reactors that do listen for reversals (or other events that do cause spam hits)? The reversal is going to be logged no matter what if it hits, so I guess there's no problem there. We'd hit the target, the period would start. The hit would be reversed, the original hit would no longer count for that period. Whether the target actually listens to reversals is irrelevant.
## Benefits

It would allow us to fire the reversal and spam actions only when a hit had actually occurred. I'm not entirely sure that we'd want to hard-code that into the API, because possibly there could be reactors that only want to listen to spam events, for example.

**Comment by JDGrimes** [comment]
## DB Schema

A big question is what our database table(s) will need to look like for this to actually be useful; what data will we need to store? If we can't make it work for both of these use cases, we have a problem. So let's see what each of them needs.

The basics would probably look something like this:

| ID | type | reactor | event | signature | datetime |
| --- | --- | --- | --- | --- | --- |

The `type` field would by `fire`, `reverse`, `spam`, etc. The signature would hold a string by which this event could be uniquely identified by the entities to which it was related. This might be a hash of a JSON encoded array with the IDs of the args as values and the slugs as keys.

### Points Reactor

For the points reactor we don't really need any other information, although it would be more efficient for us to be able to tell with a single query whether an event had already been awarded points (and was unreversed). Otherwise we have to count the number of fires and reverses/spams and calculate the difference. We have to either run two count queries or pull the rows from the database and sort and count them.

To do that, we could introduce a second table which would just hold a flag telling us if the event was awarded. However, I'm not sure that is worth it.

### Periods Extension

For the period extension we need to be able to check not only the signature of the event but the signature for the periods specifically. This is not based on the main args of the event, but on specific sub-args to which the event is tied.

We support several periods per reaction, which means that there isn't just one period per event. What we'd need to do is store this data as metadata for the fire. We could add a second table to store this information, or else add a new row to the table outlined above just to store the metadata. However, if we put this all in a row we'd need to pull the data for the fire out from the database and then check against that data, whereas we might be able to construct a query that would just join on the meta table. But it would probably be more performant not to use a join query.

**Comment by JDGrimes** [comment]
In summary of the above, this may provide a necessary part of the solution to these two problems, but perhaps it doesn't alone provide the ideal solution to either. As we work more on each of those issues, the way forward may become more clear.

**Comment by JDGrimes** [comment]
I knew that there was some scenario that might make us want to log misses, but I couldn't remember what it was. Now I do:

User publishes a post, receives points, period begins. They publish another post, period hasn't ended, they receive no points. The period ends, they go back and unpublish the post, then publish it again. They receive points.

Do we care? Most of the time users don't have the ability to toggle the post's visibility like that unless they are in a more highly trusted role. More generally, the rate limit still keeps the user from earning more than the set rate of points, so it does its job.

Also, we're currently limiting the periods extension only to repeatable events (where the status of the involved entities isn't toggled in any way). So it isn't made available in the UI to the Post Publish event.

**Comment by JDGrimes** [comment]
We might also need to keep track of misses for the rate extension (#6). Though I suppose that could be done separately when the time comes.

**Comment by JDGrimes** [comment]
> For the points reactor we don't really need any other information, although it would be more efficient for us to be able to tell with a single query whether an event had already been awarded points (and was unreversed). Otherwise we have to count the number of fires and reverses/spams and calculate the difference. We have to either run two count queries or pull the rows from the database and sort and count them.
> 
> To do that, we could introduce a second table which would just hold a flag telling us if the event was awarded. However, I'm not sure that is worth it.

Better than that, we could introduce a column into the hits log table itself. We might call it `superseded_by` or `reversed_by`, and it would contain the ID of the hit/reverse hit that came after it and reversed it. Then we could just query for the row where that column was empty. If none was found, we'd know that the hits had been reversed and so we could hit again without fear of a repeat.

**Comment by JDGrimes** [comment]
In https://github.com/WordPoints/hooks-api/issues/66#issuecomment-167806802 we realized that if we log only hits, we'll need to log them per reaction, since hits are determined on a per reaction basis. Since the reaction ID is only unique to the reactor and storage group/reaction type (standard, network, etc.) we'd need to add two columns to our table (`reaction_id` and `reaction_type`):

| ID | fire_type | reactor | reaction_id | reaction_type | event | signature | datetime |
| --- | --- | --- | --- | --- | --- | --- | --- |

**Comment by JDGrimes** [comment]
Another interesting observation from https://github.com/WordPoints/hooks-api/issues/66#issuecomment-167881161: it might make sense to store the event/fire data in one table, but the reaction-specific data in another table with a reference to that fire ID. This makes a lot of sense, since fires can cause several hits in different reactors, and duplicating the fire/event ID is unnecessary added complexity. One benefit of that which was noted there is that we can then query for the last type of fire performed for an event without having to have information on a particular reaction, which might be useful. Although I guess we could do that anyway by just querying certain rows in the table. The issue was that the different previous statuses for the event for different reactions could get desynchronized. Whether we go with this model or not will depend mostly on whether we want that to be possible.

Our tables might look something like this:

| ID | fire_type | event | signature | datetime | superseded_by |
| --- | --- | --- | --- | --- | --- |

| ID | fire_id | reactor | reaction_id | reaction_type |
| --- | --- | --- | --- | --- |

**Comment by JDGrimes** [comment]
## One tables, or Two

> The issue was that the different previous statuses for the event for different reactions could get desynchronized. Whether we go with this model or not will depend mostly on whether we want that to be possible.

This is about hits, not about fires. That's important to remember. The way I was thinking about it before was that we'd fire an event and the log would basically reflect the state of each reaction tied to that event when it fired. So the reactions would be listening for a reverse. But then a new reaction get's created that is listening for a hit. So we get to the place where some reactions are listening for reverses and some for hits. This is desynchronization, and it is actually expected. (See #68)

So from that we conclude that we don't need to have 2 tables, and it might be much better if we just had one. That'd make things less complex, although it would mean increased data storage.

However, it doesn't mean that we _have_ to have just one table, either. Because what we're logging isn't the global state of that event, but the information that a particular fire of an event had hits. The information for that fire is global to all of the individual hits, and so could still be logged in one table while the reaction-specific data was in another.

Putting the `superseded_by` column into the event hits table (as shown above) instead of the reaction hits table would imply that all hits will always be reversed by a single reverse event. But in #70 we concluded that reactors were at their own discretion when it comes to choosing whether to ignore a reversal for a changed/deleted reaction. Which would mean that the reversal event might not actually result in a real reversal taking place. On the other hand, I suppose that how the reactor decides to handle that reversal is moot; the reverse hit still happened, and the logs should reflect that. The only reason we'd want to do otherwise is if the reactor wanted to ignore this reversal but still listen for later ones. But then that doesn't make much sense either.

Instead of tying ourselves down to a table structure that is more complex and also less flexible, let's go with the simpler, single-table design. This will probably make logging a reaction hit a much simpler procedure too, because otherwise we have to check whether a hit is already recorded for the event in the one table and if not add it before recording the hit in the reaction hit table. Also, any restrictions that the two-table system might have provided can be implemented with code logic later if we change our minds about any of this.

The only potential downside I see is that we wouldn't be able to find hits from different reactions that were caused by a single event fire. I think that that is a feature which would actually be needed for #69. We need to decide that before we can proceed.

**Comment by JDGrimes** [comment]
> The only potential downside I see is that we wouldn't be able to find hits from different reactions that were caused by a single event fire. I think that that is a feature which would actually be needed for #69. We need to decide that before we can proceed.

Based on https://github.com/WordPoints/hooks-api/issues/69#issuecomment-168077422, the conclusion is that we do not actually need to be able to find hits from the same fire, just the same event generally.

But might there ever be another reason we'd need that info?

~~Think about retroactive fires. A single retroactive event can cause a whole bunch of hits. So it makes sense to be able to have a single row in the separate events table so that we can more easily undo a retroactive fire (if we ever wanted to).~~ But then that doesn't make any sense, because there would still need to be a different row for each entity/signature of the event in the event hits table. So that's a red herring.

So I don't really see any reason that we'd really need to know which hits occurred on the same fire. If we ever did need such a feature, it is pretty easy to guess based on the datetime field. So let's just start building this thing and see what happens.

But do we really need to log hits if we are going to to reversals in the points reactor? Yes, because we still need to handle refires when reversal is disabled.

**Comment by JDGrimes** [comment]
Logging all hits is necessary for us to do #68, which I think is necessary. So that means we'll be doing this.

**Comment by JDGrimes** [comment]
After #78, in which we decided that each event can have only one primary (or "signature") arg, and #77, in which we decided that signature args could not be added to or removed from an event, we might want to remove the `signature` field in favor of a `primary_arg_guid` field. I guess the name doesn't matter all that much, but it might be more obvious if we changed it to something like that. I suppose that we might still want to have the field be a hash though, since the GUID of an arg will be an array. On the other hand, perhaps we should reconsider whether this is wise—might there not be a reason why we'd want to be able to get the GUID of the primary arg from the hit logs? If we use a hash, that would be impossible.

**Comment by JDGrimes** [comment]
Also, I think we should change the name of `fire_type` to just `firer`.

**Comment by JDGrimes** [comment]
Currently the biggest hold up here is that I'm not sure what we should do with the primary arg GUIDs. Should we continue to save them as a hash value? Should we just save the serialized array? Should we not serialize them at all but somehow store the values in another table?

I guess there are two main questions that we need to answer:
1. Would we ever need to get the actual value of the entity GUID?
2. Would we ever need to query hits based on part of the entity context?

In either case, we wouldn't be able to use the hashed value. If, on the other hand, we don't need to do either of these things, storing just the hash is tempting, because it is a fixed length string that is pretty easy to query by and a short prefix of it is sufficient to use in any indexes (if we need to use any indexes).

I think that really it is too early for us to know for sure how this information might be used or queried in the future. So it would be unwise for us not to store the actual value of the entity GUID at all. However, I think that at present it would be overkill to store it in a separate table. If we find that we need to query by this a lot in the future, we can always introduce a new method of storing the value. The main thing is that we actually have the value there at all, which we wouldn't if we store only the hash.

The only question now is whether we should store both the serialized value and the hash, or just the serialized value.

After some research on hash indexes, I found [this answer on StackExchange](http://dba.stackexchange.com/a/2821/36028) that gives some useful info. Because our GUIDs are not very long strings when serialized, I suggest that we wait on trying to add a hash index. We can come back and optimize this more later if it is needed.

Edit: also, were we to do this, we could probably get [better performance using a `BINARY` field](https://mariadb.com/blog/hash-based-group-strategy-mysql) instead of `CHAR`.

**Comment by JDGrimes** [comment]
One thing which is important when we are serializing the GUID is the order of the elements in the array. If this is inconsistent, our queries won't work properly. Currently `Entity::get_the_guid()` is consistent in the order of the elements in the array that it returns, but this is not a documented feature nor is it guaranteed to remain that way in the future. The order returned is also strange: the lowest context is first in the array, then successively higher contexts follow, but the last element is the ID of the entity itself.

We'll likely want to change this, but even short of that, we may want to modify the order before serializing the data, to ensure that it is optimized for querying. I don't think it really makes much difference what order it is in though, except that when not on multisite it would probably be faster if the entity ID is first because the rest of the context will usually be the same (for entities that exist in the `site` context).

I think really that the way we currently handle contexts is too fragile as pertains to order. What we really need to do is make the GUID a multidimensional array with some sort of numeric ordering (probably just by using numeric keys). But I suppose that that is something for another ticket. For now let's just focus on removing the hashing feature.

**Comment by JDGrimes** [comment]
The only thing that we might still want to do here is add an index on the fields used in the query that retrieves the reactions to reverse. Something like:

``` sql
KEY `unsuperseded` (`event`(191), `firer`(191), `primary_arg_guid`(191), `superseded_by`)
```

**Comment by JDGrimes** [comment]
I'm going to say this is finished for now. We'll open new tickets for new issues.

The main thing that we might want to revisit in the future is optimization of the database indexes, but we'll do that once performance becomes more important.

The only other thing that we'll likely do as a follow up to this is changing the way we store the periods data.
